### ISSUE TITLE
fromFormat should prefer the IANA zone if there are multiple zone specifiers

### ISSUE THREAD

**Issue by gerob311** [moment/luxon#1076]
It currently does not seem to be possible to set a time zone on a DateTime when parsing a Java-style timestamp with a time zone.

```
describe('DateTime', () => {
    const formatString = `yyyy-MM-dd'T'HH:mm:ss.SSSZZ'['z']'`
    it('should create from timezoned string', () => {
        const time = DateTime.fromFormat('2021-11-10T13:48:00.000+08:00[Australia/Perth]', formatString, { zone: undefined });
        expect(time.zoneName).toBe('Australia/Perth');
    });
});
```

Expected: Time zone of `time` should be `Australia/Perth`
Actual: Time zone of `time` is my machine's local time zone

I'm a bit confused by this, because the documentation seems to suggest that the opts Object can override a zone specified in the parsed string, which implies to me that it should be possible to specify a zone in the parsed string. 

If I use `fromFormatExplain` instead, the time zone is listed in the `matches` object, but not the `result` object.

```
  {
      input: '2021-11-10T13:48:00.000+08:00[Australia/Perth]',
      tokens: [
        { literal: false, val: 'yyyy' },
        { literal: false, val: '-' },
        { literal: false, val: 'MM' },
        { literal: false, val: '-' },
        { literal: false, val: 'dd' },
        { literal: true, val: 'T' },
        { literal: false, val: 'HH' },
        { literal: false, val: ':' },
        { literal: false, val: 'mm' },
        { literal: false, val: ':' },
        { literal: false, val: 'ss' },
        { literal: false, val: '.' },
        { literal: false, val: 'SSS' },
        { literal: false, val: 'ZZ' },
        { literal: true, val: '[' },
        { literal: false, val: 'z' },
        { literal: true, val: ']' }
      ],
      regex: /^(\d{4})(\-)(\d{2})(\-)(\d{2})(T)(\d{2})(:)(\d{2})(:)(\d{2})(\.)(\d{3})(([+-]\d{1,2})(?::(\d{2}))?)(\[)([a-z_+-/]{1,256}?)(\])$/i,
      rawMatches: [
        '2021-11-10T13:48:00.000+08:00[Australia/Perth]',
        '2021',
        '-',
        '11',
        '-',
        '10',
        'T',
        '13',
        ':',
        '48',
        ':',
        '00',
        '.',
        '000',
        '+08:00',
        '+08',
        '00',
        '[',
        'Australia/Perth',
        ']',
        index: 0,
        input: '2021-11-10T13:48:00.000+08:00[Australia/Perth]',
        groups: undefined
      ],
      matches: {
        y: 2021,
        M: 11,
        d: 10,
        H: 13,
        m: 48,
        s: 0,
        S: 0,
        Z: 480,
        z: 'Australia/Perth'
      },
      result: {
        year: 2021,
        month: 11,
        day: 10,
        hour: 13,
        minute: 48,
        second: 0,
        millisecond: 0
      },
      zone: FixedOffsetZone { fixed: 480 }
    }

```

Perhaps I should add that there is a draft standard to extend RFC 3339, which allows this exact format (https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/). It would be nice to be able to parse timestamps represented in this way, because there are situations where converting to the local time zone is not the desired behaviour, and being able to check whether or not the zone is different to the local zone (rather than just the offset) is important.

**Comment by icambron** [comment]
There are a few things happening here:

 1. You need to pass in `{setZone: true}` to use the zone from the string as the zone setting in the datetime. Otherwise it converts it from the zone in the string to your default zone. It would be exceedingly strange if Luxon's default behavior was to set the zone based on the string, instead of simply using the offset in the string to interpret the time it represents. This is covered in the docs [here](https://moment.github.io/luxon/#/zones?id=strings-that-specify-an-offset)
 2. `{zone: undefined}` doesn't do anything
 3. Luxon doesn't have any particular rules that make it prefer the Australia/Perth to +08:00, and in practice it seems to take the former. This means that even using `setZone`, you get the fixed-offset utc+8, not the Perth zone. It also means that if the numerical offset and the IANA zone contradicted, it would use numerical offset to interpret the string whether you set `setZone` or not.

For example, this does what you expect:

```js
DateTime.fromFormat('2021-11-10T13:48:00.000[Australia/Perth]',  `yyyy-MM-dd'T'HH:mm:ss.SSS'['z']'`, {setZone: true }).
```

1 is the issue you're running into, but once you got passed that, you'd just run into 3. For 3, I'd accept a PR that makes the parser prefer to use the parsing result of `z` over the parsing result of `ZZ`. I'm going to rename this issue to reflect that that's the root problem.

> there is a draft standard to extend RFC 3339

We'll implement it when it's a standard

**Comment by gerob311** [comment]
@icambron Thanks, that's really helpful! I'm having a go at making that PR. It seems simple enough to prefer the result of `z` over the result of `ZZ`, but I'm having a couple of problems that I thought I should ask about before I go any further.

1. I found that even before I made any code changes, there were 30 failing test cases. They seem to be cases around localised formatting. For example, one of the failures is 
```
DateTime#toFormat('FF') returns a medium date/time representation with seconds

    expect(received).toBe(expected) // Object.is equality

    Expected: "May 25, 1982, 9:23:54 AM"
    Received: "25 May 1982, 9:23:54 am"
```
This makes me a little nervous about submitting a PR.

2. Now that the zone is being set, I find that if the offset conflicts with the zone, the timestamp is not converted into the specified zone. For example, the following test case fails, because the hour is still the same as the hour in the input string:
```
test("DateTime.fromFormat() converts non-matching offset when IANA zone id is present", () => {
  console.log("should be Perth");
  const i = DateTime.fromFormat("2021-11-12T09:07:13.000+11:00[Australia/Perth]", "yyyy-MM-dd'T'HH:mm:ss.SSSZZ[z]", { setZone: true } );
  expect(i.year).toBe(2021);
  expect(i.month).toBe(11);
  expect(i.day).toBe(12);
  expect(i.hour).toBe(6);
  expect(i.minute).toBe(7);
  expect(i.second).toBe(13);
  expect(i.millisecond).toBe(0);
  expect(i.offset).toBe(480)
  expect(i.zoneName).toBe("Australia/Perth")
});
```
Would it be reasonable to make this test case pass, or is that not the desired behaviour?

**Comment by icambron** [comment]
Good question. Luxon usually marks such things invalid. For example, if you provide a weekday and a year/month/day that conflict, you get an invalid DateTime.

However, I've recently soured on that approach. It makes the code more complicated without really doing much to help applications work better. So I'm disinclined to add more such logic, and I'll eventually rip out the existing stuff like that.

I'd like to think of it as A) gathering a bunch of individual facts and then B) picking among those facts to compute a date, discarding facts it doesn't need.

So I think the right thing to do is to make your proposed test pass.

**Comment by brendan-gero-humanetix** [comment]
@icambron Thanks for merging your PR! Is there any ETA on when a new release will be published?

**Comment by icambron** [comment]
@brendan-gero-humanetix released just now in 2.2.0
